text/0000-standard-lazy-types.md

```diff
@@ -0,0 +1,461 @@
+    }
+    fn from_cli(args: env::Args) -> Result<Logger, std::io::Error> { ... }
+}
+
+fn main() {
+    let logger = Logger::from_cli(env::args()).unwrap();
+
+    // Note how we use locally-created value for initialization.
+    INSTANCE.set(logger).unwrap();
+
+    // use `Logger::global()` from now on
+}
+```
+
+## Why `OnceCell` as a primitive?
+
+It is possible to imagine a type, slightly more general than `OnceCell`:
+
+```rust
+struct OnceFlipCell<U, V> { ... }
+
+impl<U, V> OnceFlipCell<U, V> {
+    const fn new(initial_value: U) -> OnceFlipCell<U, V>;
+
+    fn get_or_init<F: FnOnce(U) -> V>(&self, f: F) -> &V;
+}
+
+type OnceCell<T> = OnceFlipCell<(), T>;
+```
+
+That is, we can store some initial state in the cell and consume it during initialization.
+In practice, such flexibility seems to be rarely required.
+Even if we add a type, similar to `OnceFlipCell`, having a dedicated `OnceCell` (which *could* be implemented on top of `OnceFlipCell`) type simplifies a common use-case.
+
+## Variations of `set`
+
+The RFC proposes "obvious" signature for the `set` method:
+
+```rust
+fn set(&self, value: T) -> Result<(), T>;
+```
+
+Note, however, that `set` establishes an invariant that the cell is initialized, so a more precise signature would be
+
+```rust
+fn set(&self, value: T) -> (&T, Option<T>);
+```
+
+To be able to return a reference, `set` might need to block a thread.
+For example, if two threads call `set` concurrently, one of them needs to block while the other moves the value into the cell.
+It is possible to provide a non-blocking alternative to `set`:
+
+```rust
+fn try_set(&self, value: T) -> Result<&T, (Option<&T>, T)>
+```
+
+That is, if value is set successfully, a reference is returned.
+Otherwise, ther the cell is either fully initialized, and a reference is returned as well, or the cell is being initialized, and no valid reference exist yet.
+
+## Support for `no_std`
+
+The RFC proposes to add `cell::OnceCell` and `cell::Lazy` to `core`, while keeping `sync::OnceCell` and `sync::Lazy` `std`-only.
+However, there's a subset of `sync::OnceCell` that can be provided in `core`:
+
+```rust
+impl<T> OnceCell<T> {
+    const fn new() -> OnceCell<T>;
+    fn get(&self) -> Option<&T>;
+    fn try_set(&self, value: T) -> Result<&T, (Option<&T>, T)>
+}
+```
+
+It is possible because, while `OnceCell` needs block for full API, its internal state can be implemented as a single `AtomicUsize`, so the `core` part does not need to know about blocking.
+It is unclear if this API would be significantly useful.
+In particular, the guarantees of non-blocking `set` are pretty weak, and are not enough to implement the `Lazy` wrapper.
+
+## Poisoning
+
+As a cell can be empty or fully initialized, the proposed API does not use poisoning.
+If an initialization function panics, the cell remains uninitialized.
+An alternative would be to add poisoning, which will make all subsequent `get` calls to panic.
```

Comment:
I would like to compare the reason `Mutex` and `Once` use poisoning with `OnceCell`.

- `Mutex` is often used when multiple operations need to be done on some data to keep it in a consistent state. In that case it makes sense to poison the Mutex on panic, so other threads can know the data may now be inconsistent. It is not necessary for memory safety, but an extra safeguard.
- `std::sync::Once` can also poison itself on panic. As it guarantees to run only once, it makes sense to 'report' it when that one execution failed.

So in both cases there is a good reason to report something went wrong. Is there such a reason for `OnceCell`?
- We don't have to warn about possibly leaving things in an inconsistent state, as the `OnceCell` will simply remain empty on panic.
- The basic guarantee that makes `OnceCell` safe is that there will only ever be one write to the wrapped value. That is not necessary a guarantee that some closure will only be executed once: if it panics in one thread, another may give it another try.

So I think the question of whether poisoning on `OnceCell` and `Lazy` is desirable is tied to the question: **do we want to warn if the initialization closure is run more than once?** And how much may that warning cost?

As a guess I would say users of `OnceCell` mostly care about whether the value is initialized, and not if the initialization function is run exactly once.

So I wouldn't want to see a warning on overy use, unless you do some 'soothening'. Be it having `OnceCell.get()` returning `None`, or `Lazy` panicking on deref. That seems much to invasive.

But we could add the 'warning' with limited overhead to the `OnceCell.get_or_init()`, `OnceCell.try_get_or_init()`, and `Lazy::new()`. Just provide the closure that they take with a `bool` that indicates whether a previous attempt to initialize failed.

Example:
```rust
impl OnceCell {
    /* ... */

    pub fn get_or_init<F>(&self, f: F) -> &T
    where
        F: FnOnce(bool) -> T;
}
```
```rust
let cell = OnceCell::new();
// When you don't care about panicing, notice the extra `_`:
cell.get_or_init(|_| 21 + 21);
// If you want to handle it:
cell.get_or_init(|panicked| { if panicked { panic!(); 21 + 21 }});
```

This would be a breaking change for the current API of `once_cell`. But it seems like a nice and easy way for users to opt into stronger guarantees.

I think it would be good if the simple `OnceCell.set()` would just override a previous panic and initialize 'no matter what'.

B.t.w. the current behaviour os `lazy_static` if different from `once_cell`. `lazy_static` will poison itself on panic, and has no way to recover.

Reply:
I feel pretty strongly that, by default, we shouldn't expose tweaks for panicking behavior. Users of the API should not think about panics at all, unless they are in very specific circumstances. 

So, I think should either poison, or not poison, without any changes to the API (well, if we implement poisoning we might add some future-extension like `force_init` a-la `Once`, but that's definitely out of scope for the initial version).

I ... don't really know what would be the difference between poisoned and non-poisoned versions in practical usage. Abstract argument about the lack of invariants which could be broken favors non-poisoning. Consistency with all other std synchronization primitives favors poisoning.

Reply:
Ik think there are two more things to consider around poisoning.

#### Is it possible to build poisoning on top of `OnceCell`?

Suppose someone wants poisoning, is it possible to make a wrapper type to add it to `OnceCell`? Something like `OnceCell<Poisoner<T>>`?

- If you store the knowlegde that a previous initialization panicked inside the `OnceCell`, that would count as an initialization. There would be no way to recover / retry.
- If you store the knowlegde outside the `OnceCell`, you run into the question of [synchronization guarantees](https://github.com/matklad/rfcs/blob/std-lazy/text/0000-standard-lazy-types.md#synchronization-guarantees).

Both would add some space overhead.

#### Do we expect `lazy_static` to migrate to the implementation in the standard library?

`lazy_static` currently poisones on a panic during initialization, it would be a change in behavoir to change that. That seems like an argument to add some form of poisoning support to the standard library implementation.

@matklad I agree with you that the choice of poisoning doesn't seem the be that interesting in practice. 

How about keeping track of a 'poisoned' state in `OnceCell`, and expose a method to check it for the rare case where you want to know the initialization closure is run more than once? It can be used like:
```rust
let x: OnceCell<i32> = OnceCell::new();
x.get_or_init(|_| {
    if x.prev_init_panicked() { panic!() };
    1
});
```

This addition can also be done later as it is backwards-compatible.